### PR TITLE
Added compatibility with MDAnalysis 2.4

### PR DESCRIPTION
**Pull request by ianmkenney**
- Removed the `get_propka` function (next version must be 2.0.0)
- PropkaTraj constructor accepts atom indices for selections
- Pandas Index with a dtype of float64 instead of Float64Index, per deprecation warning from Pandas
- Removed removed tests for `get_propka` and changed the baseanalysis test names to be being the standard
- Expanded parameters for `test_single_frame_regression`
- Removed MDAnalysis version constraint in setup.py

### CODE REVIEW

**Comment by ianmkenney** [comment]
@orbeckst do you know why this linter might be failing? it does not fail for me locally and I'm using the same version of pylint...

**Comment by orbeckst** [comment]
no idea, sorry

File: propkatraj/propkatraj.py [hunk 1]
```diff
@@ -5,7 +5,7 @@
 
 from __future__ import print_function, division
 
-from six import string_types, StringIO, raise_from
+from six import StringIO, raise_from
```
Comment (orbeckst) [inline propkatraj/propkatraj.py]: We could get rid of six and just support Python >=3.8

Comment (orbeckst) [reply propkatraj/propkatraj.py]: ... but that can be another issue.

File: propkatraj/propkatraj.py [hunk 2]
```diff
@@ -152,7 +151,10 @@ class PropkaTraj(AnalysisBase):
     def __init__(self, atomgroup, select='protein', skip_failure=False,
                  **kwargs):
         if select is not None:
-            self.ag = atomgroup.select_atoms(select).atoms
+            if not isinstance(select, str):
+                self.ag = atomgroup.atoms[select].atoms
```
Comment (orbeckst) [inline propkatraj/propkatraj.py]: Why is this necessary? Can `select` be a index array??

File: propkatraj/propkatraj.py [hunk 3]
```diff
@@ -168,7 +170,7 @@ def __init__(self, atomgroup, select='protein', skip_failure=False,
 
         self.tmpfile = os.path.join(os.path.curdir, 'current.pdb')
         self.skip_failure = skip_failure
-        super(PropkaTraj, self).__init__(self.ag.universe.trajectory, **kwargs)
+        super().__init__(self.ag.universe.trajectory, **kwargs)
```
Comment (orbeckst) [inline propkatraj/propkatraj.py:183]: This looks as if this only works in Python 3. Do we properly set the Python version in setup.py and friends?

File: propkatraj/tests/test_regression.py [hunk 4]
```diff
@@ -1,6 +1,7 @@
 """
 Tests that propkatraj can be imported.
 """
+
 from __future__ import print_function
```
Comment (orbeckst) [inline propkatraj/tests/test_regression.py]: Can go if Python 3

File: propkatraj/propkatraj.py (hunk 2 above)
Comment (orbeckst) [reply propkatraj/propkatraj.py]: The doc string says that it can only be a string. If you want it to also be an index array then you need to change the docs and a versionchanged.

However, MDAnalysis analysis functions never use index arrays: either a selection or an AtomGroup. I'd probably just remove the ability to pass anything except a selection string and keep it simple.

File: propkatraj/tests/test_regression.py [hunk 5]
```diff
@@ -52,32 +53,8 @@ def pka_from_file(filename):
     return resnums, pkas
 
 
-@pytest.mark.parametrize('selection', ['protein', 'array', 'list'])
+@pytest.mark.parametrize('selection', ['protein', 'ag', 'array', 'list'])
```
Comment (orbeckst) [inline propkatraj/tests/test_regression.py]: I honestly have no idea why we are testing anything except a selection string. As far as I am concerned, I would just test "protein" and perhaps "protein and resid 1-10".

File: propkatraj/tests/test_regression.py [hunk 6]
```diff
@@ -170,56 +130,17 @@ def test_start_stop_step(tmpdir, u, start, stop, step):
         assert_almost_equal(pkas.index.values, times, decimal=5)
 
 
-@pytest.mark.parametrize('start, stop, step', [
-    (None, 5, None), (None, None, 25)
-])
```
Comment (orbeckst) [inline propkatraj/tests/test_regression.py:175]: Did you remove it because @IAlibay said that there's no need to check AnalysisBase functionality?

**Review by orbeckst — CHANGES_REQUESTED** [review]
Can we just decide to only support Python ≥ 3.8? If so, the setup.py metadata needs to be up-to-date (and we could remove `six` etc). We can then also set `MDAnalysis>=2`.

I am confused why `select` can be a list/array of indices in addition to a selection str. I can't remember when we added that but in any case, it's unlike any of the other typical MDA analysis functions so given that we do 2.0.0 #37 anyway, i suggest we remove this functionality and make it as simple as possible.

File: propkatraj/propkatraj.py (hunk 1 above)
Comment (IAlibay) [reply propkatraj/propkatraj.py]: If CI is 3.8+ then we should just make that move now.

File: .github/workflows/gh-ci.yaml [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+          last-n-minor-release: 1
+
+  main-tests:
+    if: "github.repository == 'Becksteinlab/propkatraj'"
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [macOS-latest, ubuntu-latest, windows-latest]
+        python-version: ["3.8", "3.9", "3.10"]
```
Comment (IAlibay) [inline .github/workflows/gh-ci.yaml]: Add 3.11 here?

File: .github/workflows/gh-ci.yaml [hunk 8]
```diff
@@ -0,0 +1,141 @@
+name: GH Actions Ci
```
Comment (IAlibay) [inline .github/workflows/gh-ci.yaml]: Remove the Travis file while we're at it?

File: devtools/environment.yml [hunk 9]
```diff
@@ -3,8 +3,8 @@ channels:
   - conda-forge
   - defaults
 dependencies:
-  - MDAnalysis<2.0.0
-  - MDAnalysisTests<2.0.0
+  - MDAnalysis
```
Comment (IAlibay) [inline devtools/environment.yml]: It'll at least need to be >1, but I agree with @orbeckst, we should just move to 2.0+ at this point.

File: propkatraj/propkatraj.py (hunk 2 above)
Comment (IAlibay) [reply propkatraj/propkatraj.py]: Agreed, I'm also opposed to using an index array unless there's a good reason for it.

File: propkatraj/tests/test_regression.py (hunk 5 above)
Comment (IAlibay) [reply propkatraj/tests/test_regression.py]: It seems like line 79 was erroneously removed and this one (which was intended for the old method) was kept instead.

File: propkatraj/tests/test_regression.py (hunk 6 above)
Comment (IAlibay) [reply propkatraj/tests/test_regression.py:175]: This shouldn't get removed, the def_start_stop_step above should instead.

File: propkatraj/tests/test_regression.py [hunk 10]
```diff
@@ -275,13 +178,4 @@ def test_nonprotein_residues(tmpdir):
...
-            pkas = propkatraj.get_propka(u, sel="not protein", stop=1)
-
-
-def test_nonprotein_residues_analysisbase(tmpdir):
-    """Test for Issue #24 - remove once PDBWriter is fixed"""
-    u = mda.Universe(GRO, XTC)
-    with tmpdir.as_cwd():
-        wmsg = "Non protein atoms passed"
-        with pytest.warns(UserWarning, match=wmsg):
-            pkatraj = propkatraj.PropkaTraj(u, select="not protein")
```
Comment (IAlibay) [inline propkatraj/tests/test_regression.py:287]: Why is this line being changed?

**Review by IAlibay — CHANGES_REQUESTED** [review]
I'll have a look at the pylint failure later today

File: .github/workflows/gh-ci.yaml (hunk 7 above)
Comment (ianmkenney) [reply .github/workflows/gh-ci.yaml]: Might as well!

File: devtools/environment.yml (hunk 9 above)
Comment (ianmkenney) [reply devtools/environment.yml]: @IAlibay so `>2.0.0` is fine?

File: propkatraj/tests/test_regression.py (hunk 5 above)
Comment (ianmkenney) [reply propkatraj/tests/test_regression.py]: @IAlibay I figured that I'd do the union of the two parameter sets. I'll switch it to just `protein` and `ag`

File: devtools/environment.yml (hunk 9 above)
Comment (IAlibay) [reply devtools/environment.yml]: Yup

File: propkatraj/tests/test_regression.py (hunk 10 above)
Comment (ianmkenney) [reply propkatraj/tests/test_regression.py:287]: @IAlibay style enforcement, [flake8](https://www.flake8rules.com/rules/F841.html). Alternatively we could also just not assign it?

Comment (IAlibay) [reply propkatraj/tests/test_regression.py:287]: I see, I don't particularly mind either way. I generally just don't assign since it's a test for a warning, but also "whatever works" is fine here.

File: propkatraj/propkatraj.py (hunk 3 above)
Comment (ianmkenney) [reply propkatraj/propkatraj.py:183]: @orbeckst Only mention of python version I can find is an `if` statement in `setup.py`

Comment (IAlibay) [reply propkatraj/propkatraj.py:183]: @ianmkenney the setup.py needs updating to include a python_requires (see: https://github.com/MDAnalysis/mdanalysis/blob/c7c5f0a6850d7f63b14b7697134c4cf7dd860fb9/package/setup.py#L640).

The metadata will also need updating as necessary.

File: propkatraj/tests/test_regression.py (hunk 6 above)
Comment (ianmkenney) [reply propkatraj/tests/test_regression.py:175]: @IAlibay I'm renaming the tests as I go as well since the AnalysisBase is now the only supported option. The test above was changed to test the same thing as the one I removed.

**Comment by ianmkenney** [comment]
@IAlibay @orbeckst Linting failed because MDAnalysis was not installed alongside pylint, it had no idea what attributes were defined in the superclass and just assumed I wrote something invalid. This is probably going to be an issue for projects using the cookiecutter.

**Comment by IAlibay** [comment]
> @IAlibay @orbeckst Linting failed because MDAnalysis was not installed alongside pylint, it had no idea what attributes were defined in the superclass and just assumed I wrote something invalid. This is probably going to be an issue for projects using the cookiecutter.

That should be easy enough to fix, we should push a self install upstream to the cookiecutter bits.

**Comment by ianmkenney** [comment]
I raised [an issue](https://github.com/MDAnalysis/cookiecutter-mdakit/issues/47) for the cookiecutter.

File: devtools/environment.yml [hunk 11]
```diff
@@ -3,9 +3,8 @@ channels:
   - conda-forge
   - defaults
 dependencies:
-  - MDAnalysis<2.0.0
-  - MDAnalysisTests<2.0.0
-  - six
+  - MDAnalysis>2.0.0
+  - MDAnalysisTests>2.0.0
```
Comment (orbeckst) [inline devtools/environment.yml]: Shouldn't this be `>=` because any 2.x should work

File: propkatraj/tests/test_regression.py [hunk 12]
```diff
@@ -52,46 +52,19 @@ def pka_from_file(filename):
...
-        elif selection == 'array':
-            selection = u.select_atoms('protein').ix
-            pkatraj = propkatraj.PropkaTraj(u, select=selection)
         else:
             pkatraj = propkatraj.PropkaTraj(u)
 
         pkatraj.run(stop=1)
 
-        # load reference data
+        # load reference data, makes testing arbitrary selections difficult
```
Comment (orbeckst) [inline propkatraj/tests/test_regression.py]: True, you'd probably have to slice the reference data based on the `selection.residues`. Even then, you might get differences because propka might be missing interactions. Definitely leave the updated comment!

File: setup.py [hunk 13]
```diff
@@ -36,16 +30,17 @@
                    'Programming Language :: Python :: 3.6',
                    'Programming Language :: Python :: 3.7',
                    'Programming Language :: Python :: 3.8',
```
Comment (orbeckst) [inline setup.py:26]: Can we shift the Python versions to the 3.8 - 3.11 range?

Also, these metadata have to be updated because below `python_requires='>=3.8',`

Do we test all these python versions?

File: setup.py [hunk 14]
```diff
@@ -36,16 +30,17 @@
...
-      },
+          },
       packages=['propkatraj'],
       scripts=[],
       license='GPLv3',
       long_description=open('README.md').read(),
       long_description_content_type='text/markdown; variant=GFM',
-      install_requires=['six', 'numpy', 'pandas', 'MDAnalysis<2.0.0', 'propka==3.1']
+      python_requires='>=3.8',
+      install_requires=['numpy', 'pandas', 'MDAnalysis>2.0.0', 'propka==3.1']
```
Comment (orbeckst) [inline setup.py]: `'MDAnalysis>=2.0.0'`

**Review by orbeckst — CHANGES_REQUESTED** [review]
Looking pretty good, just some issues surrounding MDA and Python versions.

File: setup.py (hunk 13 above)
Comment (ianmkenney) [reply setup.py:26]: we do not, missed those lines, mkaing it 3.8-3.11

File: devtools/environment.yml (hunk 11 above)
Comment (ianmkenney) [reply devtools/environment.yml]: It should

**Review by orbeckst — APPROVED** [review]

**Comment by ianmkenney** [comment]
@IAlibay Is the current state of this good for you?

**Comment by IAlibay** [comment]
> @IAlibay Is the current state of this good for you?

Sorry I've not been able to do much today (been mostly bed bound), I'm going to review tomorrow.

**Comment by ianmkenney** [comment]
@orbeckst @IAlibay Is there a reason we are pinning to propka 3.1 given that 3.4 is out?

File: .github/workflows/gh-ci.yaml [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+        which python
+        which pip
+        pip list
+        conda info
+        conda list
+    - name: Run tests
+      run: |
+        pytest -v --color=yes propkatraj/tests/
+
+
```
Comment (IAlibay) [inline .github/workflows/gh-ci.yaml]:
```suggestion
```
just a consistency thing

File: .github/workflows/gh-ci.yaml [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+      run: |
+        which pip
+        which python
+        pip install pylint MDAnalysis>=2.0.0
+    - name: Run Pylint
+      env:
+        PYLINTRC: .pylintrc
+      run: |
+        pylint propkatraj
+  pypi_check:
```
Comment (IAlibay) [inline .github/workflows/gh-ci.yaml:117]:
```suggestion

  pypi_check:
```
as above, empty line consistency

File: propkatraj/propkatraj.py [hunk 17]
```diff
@@ -225,146 +222,6 @@ def _conclude(self):
...
-        logging.warning('percent failure = {0:.3f}%'.format(
-            float(failed_frames)/len(universe.trajectory)*100))
-        logging.warning('failed frames: %r', failed_frames_log)
-
-    # a `pandas.DataFrame` is a good data structure for this data
-    df = pd.DataFrame(pkas, index=pd.Float64Index(times, name='time'),
-                      columns=[g.atom.resNumb for g in groups])
-
-    return df
+        self.pkas = pd.DataFrame(self._pkas, index=pd.Index(self.times,
```
Comment (IAlibay) [inline propkatraj/propkatraj.py]: If we're using MDA>2.0.0 and doing an API break, then let's also use the latest analysisbase features - could you switch to using the `results` dict here please?

File: propkatraj/tests/test_regression.py [hunk 18]
```diff
@@ -52,46 +52,19 @@ def pka_from_file(filename):
...
-        assert_almost_equal(pkas.values[0], ref_pkas, decimal=2)
-
-
-@pytest.mark.parametrize('selection', ['protein', 'ag'])
-def test_single_frame_regression_analysisbase(tmpdir, u, selection):
     """Single frame PropkaTraj regression test"""
     with tmpdir.as_cwd():
-        if selection == 'ag':
-            pkatraj = propkatraj.PropkaTraj(u.select_atoms('protein'),
+        if selection != 'default':
```
Comment (IAlibay) [inline propkatraj/tests/test_regression.py]: I can understand removing array since it wasn't in the parametrize call, but I don't necessairly  understand the need to change the rest. Maybe I'm missing the value it provides?

File: propkatraj/tests/test_regression.py (hunk 6 above)
Comment (IAlibay) [reply propkatraj/tests/test_regression.py:175]: It leads to a very confusing diff unfortunately, but yeah ok that makes sense :(

**Review by IAlibay — CHANGES_REQUESTED** [review]
Couple of things, also the license classifier doesn't match the license we have on the package itself - @orbeckst could you clarify what was the original license intent?

**Comment by orbeckst** [comment]
@IAlibay the intent with the license was "whatever MDAnalysis has". So it should be GPLv2+.

**Comment by IAlibay** [comment]
> @IAlibay the intent with the license was "whatever MDAnalysis has". So it should be GPLv2+.

Ok, we then just need to update the main license and the setup.py identifier.

File: propkatraj/tests/test_regression.py (hunk 18 above)
Comment (ianmkenney) [reply propkatraj/tests/test_regression.py]: @IAlibay This might be better.  This way we can feed in new selections easily as long as they aren't `'default'` or `'ag'`.

Comment (IAlibay) [reply propkatraj/tests/test_regression.py]: So one way to re-phrase my comment is: "are we considering a case where we might have some other way of selecting 'protein' that isn't the approach used here"?

Any other regression test would fall within a different test. Hence my inability to be convinced about the necessity of further changes here, unless there's a selection method I'm not considering?

Comment (ianmkenney) [reply propkatraj/tests/test_regression.py]: Oh, I see. If there is no desire to expand the tests to other selections then this code doesn't add anything new. I can revert it.

File: propkatraj/propkatraj.py [hunk 19]
```diff
@@ -210,161 +208,22 @@ def _single_frame(self):
...
-        logging.warning('percent failure = {0:.3f}%'.format(
-            float(failed_frames)/len(universe.trajectory)*100))
-        logging.warning('failed frames: %r', failed_frames_log)
-
-    # a `pandas.DataFrame` is a good data structure for this data
-    df = pd.DataFrame(pkas, index=pd.Float64Index(times, name='time'),
-                      columns=[g.atom.resNumb for g in groups])
-
-    return df
+        self.results.pkas = pd.DataFrame(self._pkas, index=pd.Index(self.times,
```
Comment (orbeckst) [inline propkatraj/propkatraj.py:241]: update docs that pkas are in `results.pkas`

**Review by orbeckst — CHANGES_REQUESTED** [review]
WIth use of `.results` (which is good), the docs also need to be changed
- doc string
- example notebook

File: propkatraj/propkatraj.py (hunk 19 above)
Comment (ianmkenney) [reply propkatraj/propkatraj.py:241]: done

File: .github/workflows/gh-ci.yaml (hunk 16 above)
Comment (IAlibay) [reply .github/workflows/gh-ci.yaml:117]: There's some empty line consistency issues all over this yaml 🤣 it's not technically incorrect so I won't block over it, but I don't know how I'm the only one feeling anxious looking at the formatting..

File: propkatraj/propkatraj.py [hunk 20]
```diff
@@ -210,161 +208,22 @@ def _single_frame(self):
 
     def _conclude(self):
         # Ouput failed frames
-        if self.num_failed_frames > 0:
-            perc_failure = (self.num_failed_frames / self.n_frames) * 100
+        if self.results.num_failed_frames > 0:
+            perc_failure = (self.results.num_failed_frames / self.n_frames) \
+                           * 100
```
Comment (IAlibay) [inline propkatraj/propkatraj.py]: PEP8 implicit line continuation vs explicit?

Tbh here please do just break PEP8 and put it on one line, line comprehension >>> adhering to 79 chars

File: setup.py [hunk 21]
```diff
@@ -25,27 +19,25 @@
       classifiers=['Development Status :: 6 - Mature',
                    'Environment :: Console',
                    'Intended Audience :: Science/Research',
-                   'License :: OSI Approved :: GNU General Public License v2 (GPLv2)',
+                   'License :: OSI Approved :: GNU General Public License v2+ (GPLv2+)',
```
Comment (IAlibay) [inline setup.py]:
```suggestion
                   'License :: OSI Approved :: GNU General Public License v2 or later (GPLv2+)',
```
See: https://github.com/pypa/trove-classifiers/blob/main/src/trove_classifiers/__init__.py#L276

File: propkatraj/propkatraj.py [hunk 22]
```diff
@@ -168,13 +178,14 @@ def __init__(self, atomgroup, select='protein', skip_failure=False,
...
+        super().__init__(self.ag.universe.trajectory, **kwargs)
 
     def _prepare(self):
         self._pkas = []
-        self.num_failed_frames = 0
-        self.failed_frames_log = []
-        self.failed_times = []
+        self.results.pkas = []
+        self.results.num_failed_frames = 0
+        self.results.failed_frames_log = []
```
Comment (orbeckst) [inline propkatraj/propkatraj.py:189]: Can we remove failed_frames_log and just use failed_frames?

File: setup.py [hunk 23]
```diff
@@ -25,27 +19,25 @@
...
-      },
+          },
       packages=['propkatraj'],
       scripts=[],
       license='GPLv3',
       long_description=open('README.md').read(),
       long_description_content_type='text/markdown; variant=GFM',
-      install_requires=['six', 'numpy', 'pandas', 'MDAnalysis<2.0.0', 'propka==3.1']
+      python_requires='>=3.8',
+      install_requires=['numpy', 'pandas', 'MDAnalysis>=2.0.0', 'propka==3.1']
```
Comment (IAlibay) [inline setup.py:42]: re: propka3.1, I can't fully remember. My suggestion is that after this PR we open a separate one that unpins and see what happens.

If I remember correctly 3.2 is broken, then we fixed things upstream and 3.3 works, see: https://github.com/Becksteinlab/propkatraj/issues/20. Probably in the next PR just try to pin directly to 3.3 and see if that runs fine, then we can do >=3.3?

File: propkatraj/propkatraj.py [hunk 24]
```diff
@@ -189,12 +200,12 @@ def _single_frame(self):
...
             errmsg = "failure on frame: {0}".format(self._ts.frame)
             if not self.skip_failure:
-                raise_from(RuntimeError(errmsg), err)
+                raise RuntimeError(errmsg) from err
             else:
                 warnings.warn(errmsg)
-                self.num_failed_frames += 1
-                self.failed_frames_log.append(self._ts.frame)
-                self.failed_times.append(self._ts.time)
+                self.results.num_failed_frames += 1
```
Comment (orbeckst) [inline propkatraj/propkatraj.py:208]: superfluous?

Can we remove failed_frames_log and just use failed_frames?

File: propkatraj/propkatraj.py [hunk 25]
```diff
@@ -189,12 +200,12 @@ def _single_frame(self):
...
+                self.results.failed_frames_log.append(self._ts.frame)
```
Comment (orbeckst) [inline propkatraj/propkatraj.py:209]: superfluous?

Can we remove failed_frames_log and just use failed_frames?

File: propkatraj/propkatraj.py (hunk 22 above)
Comment (orbeckst) [reply propkatraj/propkatraj.py:189]: oops misunderstood:frames vs times, will fix

File: propkatraj/propkatraj.py (hunk 25 above)
Comment (orbeckst) [reply propkatraj/propkatraj.py:209]: my bad, different things

**Comment by orbeckst** [comment]
@ianmkenney sorry, lots of doc updates, please squash eventually or just squash the whole PR

**Review by orbeckst — APPROVED** [review]

**Comment by ianmkenney** [comment]
@IAlibay do the proposed changes work for you?

**Review by IAlibay — CHANGES_REQUESTED** [review]
Sorry I thought I had pushed the request changes a couple hours ago, but for some reason it's not done it...

File: propkatraj/propkatraj.py (hunk 20 above)
Comment (ianmkenney) [reply propkatraj/propkatraj.py]: I'm typically on board with that, I don't particularly like the column maximum requirement, especially for an indentation based language. I will gladly switch it back.

File: setup.py (hunk 23 above)
Comment (ianmkenney) [reply setup.py:42]: That sounds reasonable.

**Review by IAlibay — APPROVED** [review]
lgtm!

**Comment by IAlibay** [comment]
Are we happy to squash merge?

**Comment by ianmkenney** [comment]
> Are we happy to squash merge?

Squash away
